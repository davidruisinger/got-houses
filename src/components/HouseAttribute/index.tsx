import React, { FunctionComponent, ReactChild } from 'react'
import posed from 'react-pose'
import styled from '../_utils/styledComponents'
import FlexBox from '../FlexBox'
import Container from '../Container'

const animationProps = {
  enter: {
    x: 0,
    opacity: 1,
    delay: ({ i }: { i: number }) => i * 100,
    beforeChildren: true,
  },
  exit: { x: 50, opacity: 0 },
}

const AnimatedWrapper = posed(FlexBox)(animationProps)

const Icon = styled.div`
  font-size: 2.8em;
  padding: 0.2em;

  color: ${props => props.theme.text_default};
  background-color: ${props => props.theme.canvas_ground};
`

interface Props {
  icon: ReactChild
  data: string
  i?: number
}

const HouseAttribute: FunctionComponent<Props> = ({ icon, data, i }) => (
  <AnimatedWrapper i={i} justifyContent="space-between" alignItems="stretch">
    <Icon>{icon}</Icon>
    <Container>{data}</Container>
  </AnimatedWrapper>
)

HouseAttribute.defaultProps = {
  i: 1,
}

export default HouseAttribute
