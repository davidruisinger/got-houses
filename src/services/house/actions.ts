import { createAction } from 'typesafe-actions'
import { House } from './types'

/**
 * Redux action to fetch houses.
 * @param {Object} config
 * @param {number} config.page - The page that should be fetched
 * @param {number} config.pageSize - The size of one page
 */
export const fetchHouses = createAction('GET_HOUSES_REQUEST', resolve => {
  return ({ pageSize = 10 }: { pageSize: number }) => resolve({ pageSize })
})

/**
 * Redux action to fetch a single house by ID.
 * @param {Object} config
 * @param {number} config.id - The ID of the house
 */
export const fetchHouseById = createAction(
  'GET_HOUSE_BY_ID_REQUEST',
  resolve => {
    return ({ id }: { id: string }) => resolve({ id })
  }
)

/**
 * Redux action to process an array of fetched houses.
 * @param {Object} result
 * @param {string} result.houses - The fetched houses.
 */
export const fetchHousesSuccess = createAction(
  'GET_HOUSES_SUCCESS',
  resolve => {
    return ({
      houses,
      isMoreAvailable,
    }: {
      houses: House[]
      isMoreAvailable: boolean | 'skip' // If skip is provided the reducer will ignore this (used for single fetches)
    }) => resolve({ houses, isMoreAvailable })
  }
)

/**
 * Redux action to process an error while fetching houses.
 * @param {Object} error
 * @param {string} error.message - The error message.
 */
export const fetchHousesFail = createAction('GET_HOUSES_FAIL', resolve => {
  return ({ message }: { message: string }) => resolve({ message })
})
