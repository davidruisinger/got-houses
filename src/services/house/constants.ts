export const SERVICE_NAME = 'house_service'
export const BASE_URL = 'https://anapioficeandfire.com/api/houses'
