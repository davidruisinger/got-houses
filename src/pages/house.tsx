import React, { PureComponent } from 'react'
import { RouteComponentProps } from 'react-router'
import { Dispatch } from 'redux'
import { connect } from 'react-redux'
import { State } from 'ServicesTypes'
import { House as HouseType } from '../services/house/types'
import {
  getHouse,
  getIsHousesFetching,
  getHousesError,
} from '../services/house/selectors'
import { fetchHouseById } from '../services/house/actions'
import { Layout, Spinner, HouseDetails, HouseItem } from '../components'

interface StateProps {
  error: string | null
  house?: HouseType
  getHouse: ({ id }: { id: string }) => HouseType
  isFetching: boolean
}

interface DispatchProps {
  fetchHouse: ({ id }: { id: string }) => void
}

type RouterProps = RouteComponentProps<{ id: string }>

class House extends PureComponent<StateProps & DispatchProps & RouterProps> {
  componentDidMount() {
    const { house, fetchHouse, match } = this.props
    // Fetch the house if it is not already in store
    if (!house) {
      fetchHouse({ id: match.params.id })
    } else {
      // If we do already have a house we can directly fetch the overlords
      this.fetchOverlord()
    }
  }

  componentDidUpdate(previousProps: StateProps & DispatchProps & RouterProps) {
    const { house } = this.props
    // Check if we received a house afterwards (house was fetched in componendDidMoun)
    if (house && !previousProps.house) {
      // Let's also fetch the overlords
      this.fetchOverlord()
    }
  }

  fetchOverlord() {
    const { house, fetchHouse } = this.props
    if (house && house.overlordId) {
      fetchHouse({ id: house.overlordId })
    }
  }

  handleHouseClick = (id: string) => {
    const { history } = this.props
    history.push(`/houses/${id}`)
  }

  render() {
    const { error, isFetching, house, getHouse } = this.props

    const overlord =
      house && house.overlordId && getHouse({ id: house.overlordId })

    return isFetching ? (
      <Layout title="...">
        <Spinner />
      </Layout>
    ) : house ? (
      <Layout title={house.name}>
        <HouseDetails house={house} />
        {overlord && (
          <HouseItem house={overlord} onClick={this.handleHouseClick} />
        )}
      </Layout>
    ) : (
      <Layout title="Error">
        <h1>Error:</h1>
        <h3>{error}</h3>
      </Layout>
    )
  }
}

const mapStateToProps = (state: State, ownProps: RouterProps) => ({
  house: getHouse(state, { id: ownProps.match.params.id }),
  getHouse: ({ id }: { id: string }) => getHouse(state, { id }),
  isFetching: getIsHousesFetching(state),
  error: getHousesError(state),
})

const mapDispatchToProps = (dispatch: Dispatch, ownProps: RouterProps) => ({
  fetchHouse: ({ id }: { id: string }) => dispatch(fetchHouseById({ id })),
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(House)
