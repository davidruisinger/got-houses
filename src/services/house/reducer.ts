import { ActionType, getType } from 'typesafe-actions'
import * as actions from './actions'
import { House } from './types'

export type Houses = { [id: string]: House }

export type HouseState = Readonly<{
  houses: Houses
  isFetching: boolean
  isMoreAvailable: boolean
  error: null | string
}>

const initialState: HouseState = {
  houses: {},
  isFetching: false,
  isMoreAvailable: true,
  error: null,
}

export default (
  state: HouseState = initialState,
  action: ActionType<typeof actions>
): HouseState => {
  switch (action.type) {
    case getType(actions.fetchHouses):
    case getType(actions.fetchHouseById):
      return {
        ...state,
        isFetching: true,
        error: null,
      }

    case getType(actions.fetchHousesSuccess):
      // Normalize the array to an object
      const normalizedHouses: {
        [id: string]: House
      } = action.payload.houses.reduce((obj: Houses, house) => {
        // There is no ID field so we need to get the ID from the url
        const id = house.url.split('/').pop()
        // We also transform the URL for the potentially set overlord into an ID
        const overlordId = house.overlord.split('/').pop() || ''
        if (id) obj[id] = { ...house, id, overlordId }

        return obj
      }, {})
      const isMoreAvailable = action.payload.isMoreAvailable

      return {
        ...state,
        houses: {
          ...state.houses,
          ...normalizedHouses,
        },
        isMoreAvailable:
          isMoreAvailable == 'skip' ? state.isMoreAvailable : isMoreAvailable,
        isFetching: false,
      }

    case getType(actions.fetchHousesFail):
      return {
        ...state,
        isFetching: false,
        error: action.payload.message,
      }

    default:
      return state
  }
}
