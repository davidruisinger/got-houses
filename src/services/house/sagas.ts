import { getType } from 'typesafe-actions'
import {
  put,
  takeLatest,
  call,
  select,
  delay,
  takeEvery,
} from 'redux-saga/effects'
import { BASE_URL } from './constants'
import { House } from './types'
import { getAllHouses } from './selectors'
import * as actions from './actions'

export function* fetchHousesSaga(
  action: ReturnType<typeof actions.fetchHouses>
) {
  try {
    yield delay(2000)
    const { pageSize } = action.payload
    // Since this method should only allow continiously fetch more and more houses,
    // we only need the pageSize to be passed in.
    // The actual page that is required to fetch the next products
    // is calculated based on the existing items in the store
    const existingHouses = yield select(getAllHouses)
    // NOTE: It's okay if there is an overlap where we would fetch some houses that are
    // already in the store since the reducer prevents having duplicates
    const page = Math.floor(existingHouses.length / pageSize) + 1
    const response: Response = yield call(
      fetch,
      `${BASE_URL}?page=${page}&pageSize=${pageSize}`
    )
    if (!response.ok) {
      throw new Error(response.statusText)
    }
    const json: House[] = yield call([response, response.json])
    const isMoreAvailable =
      (response.headers.get('link') || '').indexOf('rel="next"') !== -1
    yield put(actions.fetchHousesSuccess({ houses: json, isMoreAvailable }))
  } catch (e) {
    yield put(
      actions.fetchHousesFail({
        message: e.message || 'Unknown error',
      })
    )
  }
}

export function* fetchHouseByIdSaga(
  action: ReturnType<typeof actions.fetchHouseById>
) {
  try {
    yield delay(2000)
    const { id } = action.payload
    const response: Response = yield call(fetch, `${BASE_URL}/${id}`)
    if (!response.ok) {
      throw new Error(response.statusText)
    }
    const json: House = yield call([response, response.json])
    yield put(
      actions.fetchHousesSuccess({ houses: [json], isMoreAvailable: 'skip' })
    )
  } catch (e) {
    yield put(
      actions.fetchHousesFail({
        message: e.message || 'Unknown error',
      })
    )
  }
}

export default [
  takeLatest(getType(actions.fetchHouses), fetchHousesSaga),
  takeEvery(getType(actions.fetchHouseById), fetchHouseByIdSaga),
]
