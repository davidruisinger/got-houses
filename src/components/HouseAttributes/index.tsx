import React, { FunctionComponent } from 'react'
import {
  GiCrossedSwords,
  GiVerticalBanner,
  GiStoneThrone,
  GiTalk,
  GiBackup,
} from 'react-icons/gi'
import { House } from '../../services/house/types'
import HouseAttribute from '../HouseAttribute'

const attributes = [
  {
    key: 'swornMembers',
    get: (house: House) => house.swornMembers.length.toString(),
    icon: <GiBackup />,
  },
  {
    key: 'words',
    get: (house: House) => house.words,
    icon: <GiTalk />,
  },
  {
    key: 'coatOfArms',
    get: (house: House) => house.coatOfArms,
    icon: <GiVerticalBanner />,
  },
  {
    key: 'seats',
    get: (house: House) => house.seats[0] && house.seats.join(', '),
    icon: <GiStoneThrone />,
  },
  {
    key: 'ancestralWeapons',
    get: (house: House) =>
      house.ancestralWeapons[0] && house.ancestralWeapons.join(', '),
    icon: <GiCrossedSwords />,
  },
]

interface Props {
  house: House
}

const HouseAttributes: FunctionComponent<Props> = ({ house }) => {
  return (
    <>
      {attributes.map((attribute, i) => {
        const data = attribute.get(house)
        return (
          data && (
            <HouseAttribute
              icon={attribute.icon}
              data={data}
              i={i}
              key={attribute.key}
            />
          )
        )
      })}
    </>
  )
}

export default HouseAttributes
