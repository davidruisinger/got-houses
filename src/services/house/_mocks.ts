import { House } from './types'
import { HouseState } from './reducer'

export const housesMock: House[] = [
  {
    id: '1',
    url: 'https://www.anapioficeandfire.com/api/houses/1',
    name: 'Test House',
    region: 'Test Region',
    coatOfArms: 'Something',
    words: 'foo baar',
    titles: ['The Testers'],
    seats: ['Test seat'],
    currentLord: 'https://www.anapioficeandfire.com/api/characters/1',
    heir: 'https://www.anapioficeandfire.com/api/characters/1',
    overlord: 'https://www.anapioficeandfire.com/api/characters/1',
    overlordId: '1',
    founded: '',
    founder: 'https://www.anapioficeandfire.com/api/characters/1',
    diedOut: '',
    ancestralWeapons: ['Test Sword'],
    cadetBranches: [],
    swornMembers: ['https://www.anapioficeandfire.com/api/characters/1'],
  },
]

export const houseStateMock: HouseState = {
  isFetching: false,
  error: null,
  isMoreAvailable: true,
  houses: {
    [housesMock[0].name]: housesMock[0],
  },
}
