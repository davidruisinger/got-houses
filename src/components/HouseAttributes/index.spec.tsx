import 'jest-styled-components'
import React from 'react'
import { shallow, mount } from 'enzyme'
import HouseAttributes from './index'
import HouseAttribute from '../HouseAttribute'
import { housesMock } from '../../services/house/_mocks'

describe('<HouseAttributes />', () => {
  const shallowedWraper = shallow(<HouseAttributes house={housesMock[0]} />)
  const mountedWrapper = mount(<HouseAttributes house={housesMock[0]} />)

  it('matches the snapshot', () => {
    expect(shallowedWraper).toMatchSnapshot()
  })

  it('renders 5 HouseAttribute components with correct props', () => {
    const attributeComponents = mountedWrapper.find(HouseAttribute)
    expect(attributeComponents.length).toBe(5)

    expect(attributeComponents.get(0).props.data).toEqual(
      housesMock[0].swornMembers.length.toString()
    )
    expect(attributeComponents.get(1).props.data).toEqual(housesMock[0].words)
    expect(attributeComponents.get(2).props.data).toEqual(
      housesMock[0].coatOfArms
    )
    expect(attributeComponents.get(3).props.data).toEqual(
      housesMock[0].seats.join(', ')
    )
    expect(attributeComponents.get(4).props.data).toEqual(
      housesMock[0].ancestralWeapons.join(', ')
    )
  })
})
