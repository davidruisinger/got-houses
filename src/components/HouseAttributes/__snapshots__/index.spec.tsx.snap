// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HouseAttributes /> matches the snapshot 1`] = `
<Fragment>
  <HouseAttribute
    data="1"
    i={0}
    icon={<GiBackup />}
    key="swornMembers"
  />
  <HouseAttribute
    data="foo baar"
    i={1}
    icon={<GiTalk />}
    key="words"
  />
  <HouseAttribute
    data="Something"
    i={2}
    icon={<GiVerticalBanner />}
    key="coatOfArms"
  />
  <HouseAttribute
    data="Test seat"
    i={3}
    icon={<GiStoneThrone />}
    key="seats"
  />
  <HouseAttribute
    data="Test Sword"
    i={4}
    icon={<GiCrossedSwords />}
    key="ancestralWeapons"
  />
</Fragment>
`;
